hypothesis-python/src/hypothesis/database.py

```diff
@@ -22,12 +22,30 @@
 import warnings
 from hashlib import sha1
 
-from hypothesis.configuration import storage_directory
+from hypothesis.configuration import hypothesis_home_dir
 from hypothesis.errors import HypothesisException, HypothesisWarning
 from hypothesis.internal.compat import hbytes
 from hypothesis.utils.conventions import not_set
 
 
+def _usable_dir(path):
+    """
+    Returns True iff the desired path can be used as database path because
+    either the directory exists and can be used, or its root directory can
+    be used and we can make the directory as needed.
+    """
+    while True:
+        # This loop terminates because the topmost dir ('/' on unix) will always
+        # exist.
+        if os.path.exists(path):
+            usable = os.path.isdir(path) and os.access(
+                path, os.R_OK | os.W_OK | os.X_OK
+            )
+            return usable
+        else:
+            path = os.path.dirname(path)
```

Comment:
A slightly shorter version:

```python
while not os.path.exists(path):
    # Loop terminates because the root dir ("/" on unix) always exists
    path = os.path.dirname(path)
return os.path.isdir(path) and os.access(path, os.R_OK | os.W_OK | os.X_OK)
```